[WIP] Make chown tolerant of errors when doing single user mapping

kpelzel opened the pull request:
**Linked to this PR:** https://github.com/containers/storage/pull/358

### Problem:
When building a container as an unprivileged user and without uidmap tools, buildah will fail when chowning system files. This patch ignores those chown errors if buildah has fallen back to single user mapping mode.

### What this patch changes:
**Buildah side:** 
unshare.go checks if newuidmap and newgidmap is installed by running "which newuidmap". If uidmap tools are not installed it sets an environment variable to pass down to the child.

**containers/storage side:**
diff_unix.go passes in the hosts environment variables (including the single user mapping environment variable) to the "storage-applyLayer" command. This allows checkChownErr in idtools.go to check if it needs to ignore the chown errors.

### Notes:
I'm not completely sure if using environment variables is the best way to get this information down to checkChownErr, but it seemed to work best in terms minimal code change. I would love to hear feedback on better solutions.

giuseppe commented:
bot, add author to whitelist

giuseppe commented on pkg/unshare/unshare.go:
```diff
@@ -558,3 +586,25 @@ func ParseIDMappings(uidmap, gidmap []string) ([]idtools.IDMap, []idtools.IDMap,
 }
+
+func checkGIDMap() error {
+	cmd := exec.Command("which", "newgidmap")
```
I think we want to use `exec.LookPath` here instead of running another process

giuseppe commented on pkg/unshare/unshare.go:
```diff
@@ -558,3 +586,25 @@ func ParseIDMappings(uidmap, gidmap []string) ([]idtools.IDMap, []idtools.IDMap,
+}
+
+func checkUIDMap() error {
+	cmd := exec.Command("which", "newuidmap")
```
same here

kpelzel commented:
In the most recent commit I added a "single-user-mapping" global flag to buildah. This will set the environment variable that tells idtools to ignore chown errors. I didn't know what to name the flag so feedback is welcome.

I also moved most of the uidmap checking logic into MaybeReexecUsingUserNamespace because I thought it made the code more readable, but let me know if that's a bad idea.

rhatdan commented:
Changing to WIP flag, to make sure this does not get merged until we get the containers/storage patches merged and figure out how this should interact with containers/storage.

kpelzel commented:
Closing as this issue is resolved in https://github.com/containers/buildah/pull/1749

rhatdan commented:
@kpelzel Once we get this support into podman 1.5, we would love to have further feedback on what you would need to make podman a good solution for HPC environments.
